designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+
+```js
+config.setRuleSeverity("semi", "warn");
+```
+
+In this way, users don't need to modify an array to change a rule's severity, which is a major downside of this design when using a vanilla object structure as a config.
+
+The static methods on `Config` make it easy to create and modify configs. For example, suppose you want to create a config based on `eslint-config-standard` but would prefer `semi` be a warning instead of an error, you can do so using `Config.create()`:
+
+```js
+const standardConfig = require("eslint-config-standard");
+const { Config } = require("@eslint/config");
+
+const myConfig = Config.create(standardConfig, {
+    rules: {
+        semi: "warn"
+    }
+});
+
+exports.config = myConfig;
+```
+
+Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)
+
+Similarly, the `Config.assign()` method lets users modify their config by merging in changes from other configs, such as:
+
+```js
+const standardConfig = require("eslint-config-standard");
+const personalConfig = require("@personal/eslint-config");
+const { Config } = require("@eslint/config");
+
+const myConfig = Config.create(standardConfig, {
+    rules: {
+        semi: "warn"
+    }
+});
+
+Config.assign(myConfig, personalConfig);
+
+exports.config = myConfig;
+```
+
+This code example merges `personalConfig` into `myConfig` using the same functionality that ESLint currently uses to merge configs.
+
+#### The `RuleLoader` Class
+
+The purpose of the `RuleLoader` class is aid in loading rules to replace `--rulesdir` functionality and has this form:
+
+```js
+class RuleLoader {
+
+    // create a new instance
+    constructor() {}
+
+    // mimic `--rulesdir` loading
+    loadRulesFromDirectory(directory) {}
+}
+```
+
+In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `ruledefs` and then specify the rules from a directory, such as:
+
+```js
+const { Config, RuleLoader } = require("@eslint/config");
+const ruleLoader = new RuleLoader();
+
+exports.config = new Config({
+    ruledefs: {
+        custom: ruleLoader.loadFromDirectory("./custom-rules")
+    },
+    rules: {
+        "custom/my-rule": "error"
+    }
+});
+```
+
+The `RuleLoader#loadFromDirectory()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).
+
+### Advanced Configs
+
+Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:
```

Comment:
👍 

Maybe it would also make sense to support async functions, that would be helpful for use-cases like shown in `ignore` example where the user wants to read a file.

For example:

```js
const fs = require("fs").promises;

exports.config = async () => {
    const [ file1, file2 ] = await Promise.all([ fs.readFile('./file1'), fs.readFile('./file2') ]);
    
    return {
        files: "*.js",
        ignore: `${file1}\n${file2}`.replace(/\n/g, ";")
    };
});
```

Reply:
Can always be a async function?

Reply:
I don't believe we can do this as an async function *yet* because we would need to make a lot more of ESLint async to accomplish that. 

I do think we could add the ability to do async functions later on, though, as we figure out how to make more things async.